### DISCUSSION TITLE
PRef "Depression" GUI fluctuations in the reading.

### DISCUSSION THREAD

**Discussion by KirikauKiwi** [DIY-Flow-Bench/DIY-Flow-Bench#199] [Code Development]
Another thinking out loud post and related to the Feature request #180 
Please note I am using a different pressure sensor (3.3V range) to most of you but my thinking should still apply.

I was curious as to where the fluctuations in the depression reading were coming from. I am seeing a change in the tenths and hundredths of inH2O reading on the GUI with my pressure sensor connected to the ADS1115. I replaced the pressure sensor with a 10 kOhm slider to eliminate the pressure sensor and any possible fluctuations either from its sensitivities and or noise. I still see the same sorts of fluctuations. Further I removed the 10 kOhm slider and used a budget bench top PSU to send 3.3V to the ADS1115 the fluctuation are still persistent.

After seeing @DeeEmm post about using "Cool Term" in #184 I connected that up to get the Pref voltages from the API, see the videos below.

**With 10 kOhm slider**

https://github.com/user-attachments/assets/69b2e170-caf6-4ee3-ab6f-71806be7a895

**With 3.3V Benchtop PSU**

https://github.com/user-attachments/assets/1fbb2047-3be5-440c-91a9-07bf711c44a1

You can see from the videos the change in the PRef voltage that it is varying by about 187 uV which is the resolution of the ADS1115 and its configuration in DIYFB. It is fantastic to seeing it be resolved so well.

It raises the thought in my mind as I don't know enough about coding, sensor noise and electronic noise floor to the level of this.
Would this be consider as noise or possibly not a significant change in the sensors output and then either way how would this usually be delt with in the code.

Would you apply filtering/smoothing to get a stable sensor reading by filtering the PRef value so it only reports a change greater than the ADS1115 bit resolution of 187 uV e.g. 200 uV or possibly 2 bits at 400 uV.
The other alternate I can think of is using the PRef value, apply the math used in the DIYFB code and then filter/smooth it. I know mathematical you would usually do the later.

I would be interested in knowing what is considered to be the best or recommended practice.

**Comment by KirikauKiwi** [comment]
The MPXV7007 has a Fullscale span of 4.0V at a Vss of 5.0V over its 14 kPa range. this equates to approximately a resolution of measurement in the region of 1.53 Pascals = 0.006148529 inH2O. when used with the ADS115 as it is configured in DIYFB.

Example if I had 100CFM at 28 inH2O and had a gain the same as the measurable resolution I would get 
New CFM = 100 √28.006148529/28
New CFM = 100.010979

Is measuring to the 1/100th of a CFM useful, is it realistically measurable/repatable given all the variables involved.

If we expect to be able to measure to say 0.5CFM change/accuracy, this is a 0.28 inH20 change if the original CFM was 100 and the new CFM is 100.5, just a little over 1/4 of an inH20.

So to calculate the voltage we need to be able to measure for this 0.28 inH2O change.
0.006148529 inH2O/ 1 bit at 0.000187V
Voltage = (0.28/0.006148529) x 0.000187
Voltage = 0.0085159 V
which is 8516.9 uV ≈ 8.5 mV

To measure this it is usual to to have the scale of measurement to be at least one order of magnitude smaller that what is being measured.
which would give a resolution of measurement of 0.85 mV = 850 uV

Currently DIYFB is measuring at about 4.5 times this.

**Comment by DeeEmm** [comment]
> It raises the thought in my mind as I don't know enough about coding, sensor noise and electronic noise floor to the level of this. Would this be consider as noise or possibly not a significant change in the sensors output and then either way how would this usually be dealt with in the code.
 
There are always two aspects to any system, accuracy and repeatability, both are a product of the resolution of the system. As you correctly state the system generally needs to be one order of magnitude more accurate than the measurement you are trying to target.

So on that basis, to target an accuracy of 1cfm,  we should be looking at a system resolution of at least 0.1cfm. To target 0.5cfm we should be looking at a system resolution of 0.05cfm.

If we talk about the noise floor of the system, we would want it to be lower than the minimum signal that we want to detect. So for example in our 0.1cfm system we would want the noise to be less than 0.1cfm and ideally a magnitude less. We would also want to try and remove noise from entering the system in the first place. So we would apply filtering to remove noise before it enters the system. i.e. electrical filtering at the signal level.

As with all digital systems the achilles heel of accuracy is usually the Analog to Digital Converter. It is amusing to read accounts from the 80's where the greatest cost for a digital manometer / flow bench was the cost of the ADAC, where a bit resolution of 12 bits would set you back something like $3000 and bench manufacturers would often overstate the accuracy of their systems. Fortunately our little ADS1115's have a 16 bit resolution for just a few dollars. How times change.

As you have observed, The resolution of the ADC is also coupled to the range of the pressure sensor. So pressure sensors with a wider range decrease the resolution of the system., as does using the 12 bit ADS1105.

The difference between using the 16 bit ADS1115 with a sensor that has a range similar to the benches expected use range of +/- 28 "/WG compared to using the 12 bit ADS1105 with a sensor with a 200kPa range is significant...

16 bit = 4v/65,536 = 0.000061v/bit
12 bit = 4v/4096 = 0.0009765v/bit

+/-7kpa = 14kpa range, which at 0.000061v/bit = 0.000854kpa resolution
+/-100kpa = 200kpa range, which at 0.0009765v/bit = 0.1953kpa resolution

Based on the target of a 0.1cfm resolution the ADS1105 with a +/-100kpa sensor is actually out of spec and cannot achieve the 0.1 cfm resolution that is needed.

The MPX7007 or similar with the ADS1115 far exceeds the target with better than a 0.001kpa resolution, which is 100x what we need based on the one order of magnitude rule.

> Would you apply filtering/smoothing to get a stable sensor reading by filtering the PRef value so it only reports a change greater than the ADS1115 bit resolution of 187 uV e.g. 200 uV or possibly 2 bits at 400 uV. The other alternate I can think of is using the PRef value, apply the math used in the DIYFB code and then filter/smooth it. I know mathematical you would usually do the later.
> 
> I would be interested in knowing what is considered to be the best or recommended practice.

Interesting question. I'm not entirely sure what is 'best practice', or even of there is a standard solution. Both pre and post filtering are possible. It is largely process dependent. Up until now I've just been considering adding one filter to the process output, but that may change based on what we find in practice. There should be no need to 'limit' the ADS output as the output resolution should naturally resolve to 1 bit changes as it is incapable of supplying anything any different.

If we consider that we have the 'mathematical' part of the process where we take the raw signals (voltage) data and then convert it into flow, we can filter the values going into the process and also filter the resultant output. We also have data from several processes that converge mathematically to form further results, ( environmental data / pressure / flow data). We can individually filter each signal at each junction of the process to sanitise each result. 

We should also consider that each result will have a larger or smaller influence on the output of the process depending on its mathematical relationship. So we can get bigger results depending on what individual values we target. This should be determined on the accuracy of the individual process signals and how fast we would reasonably expect them to change. For example, we would generally expect the environmental inputs to change a lot slower than the pressure and flow values. 

It is entirely likely that one specific signal has a larger effect on the resultant output than the rest, and if this signal has more noise or is more unstable (for whatever reason) then there is benefit in filtering this one signal.

It should also be considered that filtering also has a negative effect in that it dampens the signal, so too much filtering can have a negative effect. Also every time you manipulate the value you essentially affect the accuracy of the data. Ideally we want to try and remove noise in the first instance and then apply mathematical filters to shape the results to remove all unexpected / out of range results. This should leave a range of data that can then be further filtered to 'smooth' or  'average' the results.

**Comment by KirikauKiwi** [comment]
I was watching a video today that happened to describe exactly what I was observing and trying to explain here unwittingly. 

I did a bit of digging and what I think is happening with the small changes in the values displayed on the GUI, is that the ADS1115's 1 LSB is 187uV and what we are seeing is the uncertainty of the ADS1115 and it is bit flipping when the analogue signal is near the transition point.

<details><summary>Here is what ChatGPT has to say on the topic.</summary>

When using the **ADS1115** or any high-precision analog-to-digital converter (ADC), **bit flipping** or **glitching** can occur near the transition between two values, especially when the analog input is close to the value corresponding to a single least significant bit (LSB) of the ADC's output. This is typically a result of **quantization noise**, where the ADC is switching between two adjacent values, often due to small fluctuations or noise in the analog signal.

To **stop or reduce bit flipping** when the analog value is close to a 1 LSB transition, you can use several techniques:

### 1. **Averaging (Multiple Samples)**
   - **Take multiple samples** and average them to smooth out the fluctuations in the signal. Since the ADC output can change rapidly near the 1 LSB transition, averaging a series of samples helps reduce noise and makes the output more stable.
   - You can either use **software averaging** by taking multiple readings and averaging them in your code or take advantage of the **internal averaging features** of the ADS1115.

   For instance, you can set the ADS1115 to **continuous conversion mode** and then average multiple samples:
   - **For continuous sampling**: Take N samples over a period and compute the average.
   - Alternatively, you could enable the **data rate reduction** on the ADS1115 (i.e., slow down the sample rate) to allow the signal to settle more effectively before each conversion.

### 2. **Use a Low-Pass Filter (Hardware or Software)**
   - A **low-pass filter** (either hardware or software) can smooth out high-frequency noise, which is often the cause of bit-flipping. A simple **RC filter** (resistor-capacitor) can help by filtering out high-frequency noise and preventing it from affecting the ADC reading.
   - The cutoff frequency of the low-pass filter should be chosen so that it **removes noise** without affecting the desired signal too much. The filter will allow the ADC to see a cleaner, more stable analog signal.
   
   For a simple **RC low-pass filter**:

<p>f<sub>c</sub> = 1 / (2 &pi; R C)</p>

   where:
<p><strong>f<sub>c</sub></strong> is the cutoff frequency.</p>
<p><strong>R</strong> is the resistance.</p>
<p><strong>C</strong> is the capacitance.</p>


   The cutoff frequency should be low enough to filter out the noise but high enough to allow the signal you are measuring to pass through.

### 3. **Increase the Resolution or Average Over More Bits**
   - The **ADS1115** offers **16-bit resolution**, which is excellent for most applications, but when you're close to a single LSB change, you might want to use **more bits of resolution** by averaging over several readings.
   - If your application is time-sensitive, a simple **averaging filter** (software or hardware) can smooth out the readings and prevent the transition between two adjacent bit values from causing instability.

### 4. **Use a Hysteresis or Deadband**
   - In software, you can implement **hysteresis** or a **deadband** to avoid unnecessary toggling of values. For example:
     - If the input signal is within **1 LSB** of a transition, you could ignore changes unless the input signal exceeds that threshold by a certain margin (for example, 2 LSB or more).
     - This approach can prevent your system from reacting to small fluctuations near the decision threshold.

   For example, if the ADC reads 1023 (just below a threshold), you can set a threshold that ignores changes until the signal passes a certain margin (e.g., only update the reading if it moves by 2 LSB or more).

### 5. **Reduce the ADC's Data Rate**
   - The **ADS1115** allows you to control the **data rate** (sampling rate). By reducing the sampling rate, the converter has more time to settle and provides a more stable reading.
   - At lower data rates (e.g., 8 SPS or 16 SPS), the output will update less frequently, and the chance of bit-flipping due to small fluctuations in the signal is reduced.

### 6. **Ensure Proper Power Supply Decoupling**
   - Noise in the **power supply** can also cause erratic ADC behavior. Ensure that your **ADS1115** is properly decoupled with **capacitors** close to the device (typically 0.1µF ceramic capacitor and a larger electrolytic capacitor like 10µF to 100µF).
   - Proper grounding and layout will also minimize power supply noise that can affect ADC accuracy.

### 7. **Utilize the Comparator Function (Optional)**
   - The **ADS1115** has a **comparator function** that can be useful if you're working with thresholds. You can use this to set an alert or interrupt when the input crosses a specific threshold, which could help reduce unnecessary fluctuations in your measurement.

---

### Conclusion:

To **prevent bit flipping** or erratic changes near the 1 LSB threshold when using the **ADS1115**, you can implement one or more of the following strategies:

- **Averaging** multiple samples (either in software or using the internal averaging of the ADS1115).
- **Low-pass filtering** the analog signal to remove noise.
- Implementing **hysteresis** or a **deadband** in software to avoid reacting to small changes.
- Lowering the **sampling rate** for more stable readings.
- Ensuring good **power supply decoupling** and proper **PCB layout**.

By combining these techniques, you can achieve more stable ADC readings and minimize bit-flipping issues, especially when your analog signal is near a transition point.
</details>

**Reply by DeeEmm** [reply]
**TLDR**

> what we are seeing is the uncertainty of the ADS1115 and it is bit flipping when the analogue signal is near the transition point.

It is not uncertainty, it is the resolution of the system. The input is crossing the threshold, this is completely normal and expected. There is no need to filter down below this level. You do not increase accuracy as you cannot interpolate between adjacent values.

As already stated, the LSB for the MPX7007 with the ADS1115 has a value of 0.000854kpa.

If we are seeing a large change in the final calculated value, as a result of the LSB being worth something approaching the target resolution of the system, then the issue is that the sensor selection is incorrect. This applies to all sensors. Range and resolution is 
a major part of the sensor selection process. It is why we need to operate in orders of magnitudes higher than the target resolution of the system.

IMHO, there is little value in trying to use a a 200kPa range sensor as we only ending up using up to 7kPa of the overall range which gives us a resolution of 0.01 kPA. So that 'bit flipping' will be visible in the GUI as we are viewing the data down to 2 decimal places. This is the same reason that support for the ADS1005 was dropped, it simply does not have a high enough resolution. 

Not sure what the range of the sensor you are using, but if the recommended sensor is used, then this should be a non-issue as the fluctuations are orders of magnitude below what is reported in the GUI.


Also. Can we put Chat-GPT stuff in spoiler blocks. 

I will be completely and brutally honest here. I don't read it. It is word soup. 

(no offense meant of course) 

**Example**

<details><summary>This is what Cicero had to say</summary>
Lorem ipsum dolor sit amet, consectetur adipiscing elit. In facilisis auctor nunc vitae rhoncus. Curabitur velit massa, aliquet at mattis sit amet, cursus id mauris. Vestibulum ac viverra turpis, ac sollicitudin nisl. Donec luctus augue quis nisi ultrices malesuada. Mauris tempor aliquam facilisis. Etiam ut aliquam est, non mollis risus. Donec semper nibh est, vel suscipit ligula suscipit vel. Nullam quis eros eget magna auctor pharetra. Pellentesque ac augue sem. Praesent tempor sapien tristique, commodo elit sed, condimentum sem. Cras egestas odio consectetur, luctus elit et, tincidunt felis. Nulla quis interdum orci, nec sodales nisi..
</details>

```
<details><summary>This is what Cicero had to say</summary>
Lorem ipsum dolor sit amet, consectetur adipiscing elit. In facilisis auctor nunc vitae rhoncus. Curabitur velit massa, aliquet at mattis sit amet, cursus id mauris. Vestibulum ac viverra turpis, ac sollicitudin nisl. Donec luctus augue quis nisi ultrices malesuada. Mauris tempor aliquam facilisis. Etiam ut aliquam est, non mollis risus. Donec semper nibh est, vel suscipit ligula suscipit vel. Nullam quis eros eget magna auctor pharetra. Pellentesque ac augue sem. Praesent tempor sapien tristique, commodo elit sed, condimentum sem. Cras egestas odio consectetur, luctus elit et, tincidunt felis. Nulla quis interdum orci, nec sodales nisi..
</details>
```

I will also update the posting guidelines
